src/main/scala/com/fulcrumgenomics/util/MetricBuilder.scala

```diff
@@ -0,0 +1,165 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2022 Fulcrum Genomics
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+package com.fulcrumgenomics.util
+
+import com.fulcrumgenomics.commons.CommonsDef.{forloop, unreachable}
+import com.fulcrumgenomics.commons.reflect.{ReflectionUtil, ReflectiveBuilder}
+import com.fulcrumgenomics.commons.util.{LazyLogging, Row}
+
+import java.io.{PrintWriter, StringWriter}
+import scala.reflect.runtime.{universe => ru}
+import scala.util.{Failure, Success}
+
+case class MetricBuilderException private[util](message: Option[String] = None) extends Exception
+
+/** Class for building metrics of type [[T]].
+  *
+  * This is not thread-safe.
+  *
+  * @param source optionally, the source of reading (e.g. file)
+  * @tparam T the metric type
+  */
+class MetricBuilder[T <: Metric](source: Option[String] = None)(implicit tt: ru.TypeTag[T]) extends LazyLogging {
+  // The main reason why a builder is necessary is to cache some expensive reflective calls.
+  private val clazz: Class[T]   = ReflectionUtil.typeTagToClass[T]
+  private val reflectiveBuilder = new ReflectiveBuilder(clazz)
+  private val names             = Metric.names[T]
+  private val namesSet          = names.toSet
+
+  /** Builds a metric from a [[ com.fulcrumgenomics.commons.util.Row]] produced from a
+    * [[ com.fulcrumgenomics.commons.util.DelimitedDataParser]].
+    *
+    * @param row        the row to parse.  All required fields must be given.  Can be in any order.
+    * @param lineNumber optionally, the line number when building a metric from a line in a file
+    * @param ignoreExtra ignore extra fields (that are not in the metric)
```

Comment:
Missing `@param` headers FYI.

But isn't both the builder and `Row` aware of their own headers/names?

Reply:
1. There may be more headers in the input file (row) than the metric
2. Row does know, but it could: https://github.com/fulcrumgenomics/commons/issues/87

Reply:
Fixed in https://github.com/fulcrumgenomics/commons/pull/88